### PR TITLE
Supports .mcfunction optimization

### LINKED ISSUE #67: Supports .mcfunction optimization
### Easy way
- Delete empty lines
- Delete comments

### Difficult way
- Replace with a shorter command with the same behavior (but it is version dependent and does advanced processing)
- Merge multiple files (but it needs to find out if they can be merged)

### PR DESCRIPTION
**Pull request by sya-ri**
close #67

## Changes

- [x] Delete empty lines and comments (`#` or `//`)
- [x] Delete indents
- [x] Delete BOM bytes
- [x] Error starting line with `/`
- [x] Writing the correct comment (Could you do it?)

### CODE REVIEW

**Comment by AlexTMjugador** [comment]
I see CI checks were skipped due to some changes I did to accomodate Dependabot PRs. I'll fix that and review your PR in some hours if possible.

**Comment by AlexTMjugador** [comment]
I've handled the CI failures, or so I think. Now the tests on Windows error out due to test text files being checked out with Windows-style CR + LF line breaks:

Windows test errors

Would you mind adding those files to the [.gitattributes file](https://github.com/ComunidadAylas/PackSquash/blob/bf8f6e1d4000284a40c3d6a1d3b102f239c36584/.gitattributes), to make sure they are checked out as expected on Windows platforms, like it was done for other test files? :heart:

File: packages/packsquash/src/pack_file/commands_function_file.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	if line_number.is_first() && strip_bom && line.chars().next().map_or(false, |c| c == '\u{feff}') {
+		line.remove(0);
+	}
+
+	let trimmed_line = line.trim();
+
+	// Check whether the line is a comment. If so, bail out by copying or skipping
+	// it. It's copied only if we're not minifying
+	if trimmed_line.is_empty() || trimmed_line.starts_with('#') || trimmed_line.starts_with("//") {
+		(!minify).then(|| prepare_for_output(line, is_last, NOT_MINIFIED))
```
Comment (AlexTMjugador) [inline packages/packsquash/src/pack_file/commands_function_file.rs:115]: `//` is never allowed in vanilla .mcfunction files from 1.12 onwards to begin a comment line: the game rejects any line that begins with a slash, but it gives a different error message when using a double slash.

Allowing `//` as a comment sequence has the unintended side effect of PackSquash outputting lines that might invalid to the game when minification is disabled, which is a bad thing. I see two ways of handling this:

- You take into account the pack metadata information to know what Minecraft version this pack targets, and decide whether to allow `//` or not based on that. This will work fine and not reject files that are valid for older versions, but it is more involved to implement and will allow pack authors to get away with arguably obsolete syntax in their functions, which will trade effort fixing functions now for effort fixing functions later, when they decide to upgrade the pack.
- Not allowing `//` altogether. This will reject valid files for older Minecraft versions, but due to the reasons stated above it's a bad idea to begin comments like this anyway. I like this option more, as it's simpler and encourages pack authors to future-proof their packs more, which is beneficial in the long term.

Comment (AlexTMjugador) [reply packages/packsquash/src/pack_file/commands_function_file.rs:115]: How do you prefer to tackle this? :smile:

**Review by AlexTMjugador — CHANGES_REQUESTED** [review]
I had some time to properly review your PR. I'm really grateful for all of your useful contributions. Thank you :heart:

In general, your code looks pretty good. I'm proud to see you're learning Rust so fast and well, noticing all those little details that help making your PR readily mergeable, such as adding unit tests and refactoring what would become duplicate code, like the `MarkLastDecorator` struct, into the appropriate utility module. With that said, I'd still like to share and address some feedback before merging.

Also, considering that Windows tests still fail in CI due to line break character differences, please let me know if you can handle that, or if you'd prefer that I fix those failures instead.

File: .gitattributes [hunk 2]
```diff
@@ -1,6 +1,7 @@
...
 # used for comparing expected results in unit tests
-packages/packsquash/src/pack_file/json_file/*.json				eol=lf
-packages/packsquash/src/pack_file/properties_file/*.properties	eol=lf
-packages/packsquash/src/pack_file/shader_file/*.fsh				eol=lf
-packages/packsquash/src/pack_file/legacy_lang_file/*.lang		eol=lf
+packages/packsquash/src/pack_file/json_file/*.json				       eol=lf
+packages/packsquash/src/pack_file/properties_file/*.properties	       eol=lf
+packages/packsquash/src/pack_file/shader_file/*.fsh				       eol=lf
+packages/packsquash/src/pack_file/legacy_lang_file/*.lang              eol=lf
+packages/packsquash/src/pack_file/commands_function_file/*.mcfunction  eol=lf
```
[the remark this answers is not in the document]
Comment (sya-ri) [reply .gitattributes]: Fix 8a2973c304915454b559d67040f2eb2e4f70f805

File: packages/packsquash/src/pack_file/commands_function_file.rs (hunk 1 above)
Comment (sya-ri) [reply packages/packsquash/src/pack_file/commands_function_file.rs]: If you use `//`, it will give a different error than `/`. This change is in this commit: e9b895630ed55b361d2ba72b83992c4f666bf245.

File: packages/packsquash/src/pack_file/commands_function_file.rs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	// The lines codec takes care of stripping line breaks from the line, even if
+	// Windows line endings (CR + LF) are used. However, we should remove the BOM in
+	// the first line if present. This would remove the BOM prefix from the key of
+	// the first command string, avoiding pack author confusion and doing a better
+	// work than Mojang ;). This might break packs that dealt with the BOM by
+	// including it in the key references elsewhere, so only do it if the option
+	// is enabled
+	if line_number.is_first() && strip_bom && line.chars().next().map_or(false, |c| c == '\u{feff}') {
+		line.remove(0);
+	}
```
[the remark this answers is not in the document]
Comment (sya-ri) [reply packages/packsquash/src/pack_file/commands_function_file.rs]: Fix ea75c9e10b3cb8614e0a7d1538b17c6c1f23b3cb

File: packages/packsquash/src/pack_file/commands_function_file.rs [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+		let mut line_number = LineNumber::new();
+
+		let minify = self.optimization_settings.minify;
+		let strip_bom = self.optimization_settings.strip_bom;
+
+		MarkLastDecorator::new(FramedRead::new(
+			self.read,
+			// Limit line length to 16 KiB to bound memory consumption and be nice to Minecraft.
+			// Longer lines are probably an error, and will negatively affect Minecraft performance
+			LinesCodec::new_with_max_length(16 * 1024)
```
[the remark this answers is not in the document]
Comment (AlexTMjugador) [reply packages/packsquash/src/pack_file/commands_function_file.rs:51]: Command parsing is done using [Brigadier's `StringReader` class](https://github.com/Mojang/brigadier/blob/cf754c4ef654160dca946889c11941634c5db3d5/src/main/java/com/mojang/brigadier/StringReader.java), which wraps a Java `String` with a `int` cursor to read substrings from it as needed. Therefore, my previous pessimistic upper bound turned to be the length limit for a single command read from a function, assuming that Minecraft has enough memory available to hold the string.

However, actually parsing the command involves additional memory allocations, and the vanilla launcher by default assigns a maximum Java heap size from 1 to 2 GiB. This means that the ~2 GiB upper limit can't be reached in practice with the default launcher options (and a 2 GiB command will likely be so slow to parse, too). Modded environments require more RAM, but it's fair to say that this extra RAM will be used by mods in practice, and thus it will not be available for command functions. A 16 MiB limit here looks sensible:

- Given that several files can be processed in parallel by PackSquash, and that the beefiest CPUs I know have 128 threads, this will bound PackSquash memory consumption to a maximum of 2 GiB, which should be a low memory consumption for sensible computers that use this kind of CPUs.
- With a maximum heap size of 2 GiB, this would mean that Minecraft can deal with up to 128 commands this long. 128 is a very low number, considering that functions tend to consist of tens of commands at least, and that other parts of the game and command parsing code require memory too (in fact, it's reasonable to say that the game requires at least 512 MiB for other purposes, which reduces the theoretical maximum number of commands to 96). I think it's extremely unlikely that someone ever needed a command 16 MiB long.
```suggestion
			// Limit line length to 16 MiB to bound memory consumption and be nice to Minecraft.
			// Longer lines are probably an error, and will negatively affect Minecraft performance
			LinesCodec::new_with_max_length(16 * 1024 * 1024)
```

[the remark this answers is not in the document]
Comment (sya-ri) [reply packages/packsquash/src/pack_file/commands_function_file.rs]: Perfect!! Thank you for committing :heart:

**Comment by sya-ri** [comment]
I thought to fix some wording but you fixed it 😍 So ready to merge PR!!!

**Comment by AlexTMjugador** [comment]
Ignore my theory of `.gitattributes` changes not being applied due to a complicated reason: the sample data files had a `.mcfuntion` extension, not `.mcfun`**c**`tion`.
